pkg/nmi/server/server.go

```diff
@@ -463,17 +491,18 @@ func listPodIDsWithRetry(ctx context.Context, kubeClient k8s.Client, logger *log
 				for _, podID := range idStateMap[""] {
 					if strings.EqualFold(rqClientID, podID.Spec.ClientID) {
 						logger.Warningf("found assignedIDs with no state for pod:%s/%s. AssignedIDs created with old version of mic.", podns, podname)
-						return idStateMap[""], nil
+						return idStateMap[""], true, nil
 					}
 				}
 				for _, podID := range idStateMap[aadpodid.AssignedIDAssigned] {
 					if strings.EqualFold(rqClientID, podID.Spec.ClientID) {
-						return idStateMap[aadpodid.AssignedIDAssigned], nil
+						return idStateMap[aadpodid.AssignedIDAssigned], true, nil
 					}
 				}
 			}
-			if len(idStateMap[aadpodid.AssignedIDCreated]) == 0 && attempt >= maxAttemptsPerCheck {
-				return nil, fmt.Errorf("getting assigned identities for pod %s/%s in CREATED state failed after %d attempts. Error: %v", podns, podname, maxAttemptsPerCheck, err)
+			if len(idStateMap[aadpodid.AssignedIDCreated]) == 0 && attempt >= maxAttemptsForCreated {
```

Comment:
What happens if we move to assigned state before we reach this check ?

Reply:
The identities are either in assigned state or created state. So during that iteration if its in assigned state, then the check for assigned state first will take care of it. There is no intermittent state possible which will result in assigned and created being empty at same time.